### DISCUSSION TITLE
Does Inventree could fit my needs (without too much dev)

### DISCUSSION THREAD

**Discussion by Bbillyben** [inventree/InvenTree#3396] [Q&A] [answered]
Hi Everybody, 

Before digging into installation and customization and putting too much time in it, I'm trying to assess wether Inventree could fit my needs. 

We need a stock management software for our research lab.
The objective is to ease stock and destock actions for our technicians, with the help of an a hand scanner, and keep track of everything related to stock. 

Some of barcodes are standard XX digit code (as EAN-13), but some are EAN-128/GS1-128, which contain several information(EAN, batch number, expiration date, quantity, .....)

I've got a java app that handle the scanning of those barcode and decode, gather intformation we needs, and is ready to send it to the stock mngt software (by rest api for instance).

Does Inventree is able to identify product by EAN ? and add/remove product by API call with this EAN ?
(I'm pretty sure he is, this is the intro question).

Could we attach a batch number to stock entry as well as expiration date by API call (with the 'add to stock' call would be the best) ? 

And could we destock/consume a stock entry, identifyed by the XX digit barcode and the batch number ? 

I havent yet access to the 'self documented' API documentation as I stuck on some installation issue.

Thanks for any incents.

Ben

**Comment by SchrodingersGat** [comment, marked as answer]
@Bbillyben I believe that most (if not all) of your requirements are covered.

### Barcodes

InvenTree has extensive barcode (and qrcode) support "out of the box" and you can assign arbitrary barcodes to stock items (and locations, if you want). If you have a scanner which can read the barcode, and access the InvenTree API, then it should work just fine.

### Batch Codes

Batch codes are supported for stock items

### Expiry Dates

https://inventree.readthedocs.io/en/latest/stock/expiry/

**Reply by Bbillyben** [reply]
Thanks for the info, I will give a try on it as soon as I can overcome my install issue (related to https://github.com/inventree/InvenTree/issues/3388)

**Comment by Bbillyben** [comment]
Hi @SchrodingersGat , Hi all

I'm exploring the API to figure out the philosophy of InvenTree, and how I could implement api call.

So several questions come up (and momre is comming):

1. About barcodes : I understand they could be linked to stock items. 
Is there any way to link a barcode to a part ( or sub part/derived part) instead of stock item? or manufacturer/supplier part ? 
I would believe barcode are consistents amongs items for a given part, or a givent part in supplier/manufacturer, more than for a single item in stock. (It changes the workflow of product identification and management amongs the solution).

2. To by pass that (for me) limitation, I'm trying to use MPN/SKU. I'm able to search against MPN, but I h'aven't found ani api call for SKU? Do I miss it? 

3. side question (quite far upstream of any dev...) : If I would dev a barcode decoder plugin (I've read fast accross plugin overview/how to/barcode mixin docs).
In a dev environement, will the plugin be updated on code update or would have i to restart the server each time?

**Comment by SchrodingersGat** [comment]
> About barcodes : I understand they could be linked to stock items.

Currently barcodes can be associated with parts, but not supplier or manufacturer parts. That could be an added feature, and I see some benefit there.

> To by pass that (for me) limitation, I'm trying to use MPN/SKU. I'm able to search against MPN, but I h'aven't found ani api call for SKU? Do I miss it?

Parts can be searched by SKU:

https://github.com/inventree/InvenTree/blob/622c0636ef1f83646c4c935ed7677d15c2c0e578/InvenTree/part/api.py#L1397

It looks like stock items cannot currently be searched by SKU. Might be worth raising an issue about that.

> In a dev environment, will the plugin be updated on code update or would have i to restart the server each time?

You can "trick" the server into auto-reloading plugin code by explicitly *not* running the background worker process. This forces any code to run in the foreground (blocking) and means that you can quickly dev plugins with auto-code-reload

**Reply by Bbillyben** [reply]
Ok, it sounds that I should consider going for the master branch.....

SKU search seems not being available in 0.7.5, I suppose it's the same for part barcode...!

**Reply by Bbillyben** [reply]
@SchrodingersGat

I've upgraded to the master branch and run the update.
I'm now on v0.8.0dev, API V67

Could point me the syntax for both request (SKU part search and parts barcode) ? 


* For MPN I perform a` GET https://myserver/api/company/part/manufacturer/?MPN=111`

But for SKU? I've tryed : `GET https://myserver/api/company/part/supplier/?SKU=111`
or `GET https://myserver/api/part/?SKU=111` 
or `GET https://myserver/api/part/?supplier_parts__SKU=111` 

(I've seen they are not describe in the doc)

* For barcode : `POST  https://myserver/api/barcode/link/` require `stockitem `key
* 
and `PATCH https://myserver/api/part/5/` with a barcode key do not add any barcode ref to the part.

**Comment by SchrodingersGat** [comment]
> Could point me the syntax for both request (SKU 

PR adding SKU filtering for SupplierPart list is here : https://github.com/inventree/InvenTree/pull/3417

> and PATCH https://myserver/api/part/5/ with a barcode key do not add any barcode ref to the part.

While parts can support internal barcodes, currently custom barcodes are not supported. There is an open issue about this: https://github.com/inventree/InvenTree/issues/1971

**Reply by Bbillyben** [reply]
Thanks for the info! 
I haven't see it was not merged yet (at that time!)
